libcaf_net/caf/net/basp/application.hpp

```diff
@@ -75,116 +84,436 @@ class CAF_NET_EXPORT application {
 
   // -- interface functions ----------------------------------------------------
 
-  template <class Parent>
-  error init(Parent& parent) {
+  template <class LowerLayerPtr>
+  error init(socket_manager* owner, LowerLayerPtr down, const settings& cfg) {
     // Initialize member variables.
-    system_ = &parent.system();
+    owner_ = owner;
+    system_ = &owner->system();
     executor_.system_ptr(system_);
     executor_.proxy_registry_ptr(&proxies_);
-    // Allow unit tests to run the application without endpoint manager.
-    if constexpr (!std::is_base_of<test_tag, Parent>::value)
-      manager_ = &parent.manager();
-    size_t workers;
-    if (auto workers_cfg = get_if<size_t>(&system_->config(),
-                                          "caf.middleman.workers"))
-      workers = *workers_cfg;
-    else
-      workers = std::min(3u, std::thread::hardware_concurrency() / 4u) + 1;
+    max_throughput_ = get_or(cfg, "caf.scheduler.max-throughput",
+                             defaults::scheduler::max_throughput);
+    auto workers = get_or<size_t>(
+      cfg, "caf.middleman.workers",
+      std::min(3u, std::thread::hardware_concurrency() / 4u) + 1);
     for (size_t i = 0; i < workers; ++i)
       hub_->add_new_worker(*queue_, proxies_);
     // Write handshake.
-    auto hdr = parent.next_header_buffer();
-    auto payload = parent.next_payload_buffer();
-    if (auto err = generate_handshake(payload))
-      return err;
-    to_bytes(header{message_type::handshake,
-                    static_cast<uint32_t>(payload.size()), version},
-             hdr);
-    parent.write_packet(hdr, payload);
-    parent.transport().configure_read(receive_policy::exactly(header_size));
-    return none;
+    auto app_ids = get_or(cfg, "caf.middleman.app-identifiers",
+                          application::default_app_ids());
+    return write_message(down, header{message_type::handshake, version},
+                         system().node(), app_ids);
   }
 
-  error write_message(packet_writer& writer,
-                      std::unique_ptr<endpoint_manager_queue::message> ptr);
-
-  template <class Parent>
-  error handle_data(Parent& parent, byte_span bytes) {
-    static_assert(std::is_base_of<packet_writer, Parent>::value,
-                  "parent must implement packet_writer");
-    size_t next_read_size = header_size;
-    if (auto err = handle(next_read_size, parent, bytes))
-      return err;
-    parent.transport().configure_read(receive_policy::exactly(next_read_size));
-    return none;
+  template <class LowerLayerPtr>
+  bool prepare_send(LowerLayerPtr& down) {
```

Comment:
There is no benchmark that gives such distinct insight into the stack. At least not as of right now.
I used this approach, since it was used `libcaf_io` with the `basp_broker`.